simapp/mint_fn.go

```diff
@@ -0,0 +1,123 @@
+// Note that this function can not have the mint keeper as a parameter because it would create a cyclic dependency.
+func ProvideExampleMintFn(bankKeeper bankkeeper.Keeper) minttypes.MintFn {
+	return func(ctx context.Context, env appmodule.Environment, minter *minttypes.Minter, epochId string, epochNumber int64) error {
+		// in this example we ignore epochNumber as we don't care what epoch we are in, we just assume we are being called every minute.
+		if epochId != "minute" {
+			return nil
+		}
+
+		var stakingParams stakingtypes.QueryParamsResponse
+		err := env.QueryRouterService.InvokeTyped(ctx, &stakingtypes.QueryParamsRequest{}, &stakingParams)
+		if err != nil {
+			return err
+		}
+
+		var bankSupply banktypes.QuerySupplyOfResponse
+		err = env.QueryRouterService.InvokeTyped(ctx, &banktypes.QuerySupplyOfRequest{Denom: stakingParams.Params.BondDenom}, &bankSupply)
+		if err != nil {
+			return err
+		}
+		stakingTokenSupply := bankSupply.Amount
+
+		var mintParams minttypes.QueryParamsResponse
+		err = env.QueryRouterService.InvokeTyped(ctx, &minttypes.QueryParamsRequest{}, &mintParams)
+		if err != nil {
+			return err
+		}
+
+		var stakingPool stakingtypes.QueryPoolResponse
+		err = env.QueryRouterService.InvokeTyped(ctx, &stakingtypes.QueryPoolRequest{}, &stakingPool)
+		if err != nil {
+			return err
+		}
+
+		// bondedRatio
+		bondedRatio := math.LegacyNewDecFromInt(stakingPool.Pool.BondedTokens).QuoInt(stakingTokenSupply.Amount)
+		minter.Inflation = minter.NextInflationRate(mintParams.Params, bondedRatio)
+		minter.AnnualProvisions = minter.NextAnnualProvisions(mintParams.Params, stakingTokenSupply.Amount)
+
+		// because we are minting every minute, we need to divide the annual provisions by minutes in a year (525600)
+		provisionAmt := minter.AnnualProvisions.QuoInt64(525600)
+		mintedCoin := sdk.NewCoin(mintParams.Params.MintDenom, provisionAmt.TruncateInt())
+		mintedCoins := sdk.NewCoins(mintedCoin)
+		maxSupply := mintParams.Params.MaxSupply
+		totalSupply := stakingTokenSupply.Amount
+
+		// if maxSupply is not infinite, check against max_supply parameter
+		if !maxSupply.IsZero() {
+			if totalSupply.Add(mintedCoins.AmountOf(mintParams.Params.MintDenom)).GT(maxSupply) {
+				// calculate the difference between maxSupply and totalSupply
+				diff := maxSupply.Sub(totalSupply)
+				// mint the difference
+				diffCoin := sdk.NewCoin(mintParams.Params.MintDenom, diff)
+				diffCoins := sdk.NewCoins(diffCoin)
+
+				// mint coins
+				if diffCoins.Empty() {
+					// skip as no coins need to be minted
+					return nil
+				}
+
+				if err := bankKeeper.MintCoins(ctx, minttypes.ModuleName, diffCoins); err != nil {
+					return err
+				}
+				mintedCoins = diffCoins
+			}
+		}
+
+		// mint coins if maxSupply is infinite or total staking supply is less than maxSupply
+		if maxSupply.IsZero() || totalSupply.Add(mintedCoins.AmountOf(mintParams.Params.MintDenom)).LT(maxSupply) {
```

Comment:
when total + minted == max supply, then neither L67 or this condition fits and no coins would be minted although a sub amount would be possible.  I think L67 should handle this

Reply:
I'm simplifying it like this:

```go
		if !maxSupply.IsZero() {
			// supply is not infinite, check the amount to mint
			remainingSupply := maxSupply.Sub(totalSupply)

			if remainingSupply.LTE(math.ZeroInt()) {
				// max supply reached, no new tokens will be minted
				// also handles the case where totalSupply > maxSupply
				return nil
			}

			// if the amount to mint is greater than the remaining supply, mint the remaining supply
			if mintedCoin.Amount.GT(remainingSupply) {
				mintedCoin.Amount = remainingSupply
			}
		}

		if mintedCoin.Amount.IsZero() {
			// skip as no coins need to be minted
			return nil
		}

		mintedCoins := sdk.NewCoins(mintedCoin)
		if err := bankKeeper.MintCoins(ctx, minttypes.ModuleName, mintedCoins); err != nil {
			return err
		}
```